pkg/logql/first_last_over_time.go

```diff
@@ -0,0 +1,267 @@
+package logql
+
+import (
+	"math"
+	"time"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/grafana/loki/pkg/iter"
+)
+
+// newFirstWithTimestampIterator returns an iterator the returns the first value
+// of a windowed aggregation.
+func newFirstWithTimestampIterator(
+	it iter.PeekingSampleIterator,
+	selRange, step, start, end, offset int64) RangeVectorIterator {
+	inner := &batchRangeVectorIterator{
+		iter:     it,
+		step:     step,
+		end:      end,
+		selRange: selRange,
+		metrics:  map[string]labels.Labels{},
+		window:   map[string]*promql.Series{},
+		agg:      nil,
+		current:  start - step, // first loop iteration will set it to start
+		offset:   offset,
+	}
+	return &firstWithTimestampBatchRangeVectorIterator{
+		batchRangeVectorIterator: inner,
+	}
+}
+
+type firstWithTimestampBatchRangeVectorIterator struct {
+	*batchRangeVectorIterator
+	at []promql.Sample
+}
+
+func (r *firstWithTimestampBatchRangeVectorIterator) At() (int64, StepResult) {
+	if r.at == nil {
+		r.at = make([]promql.Sample, 0, len(r.window))
+	}
+	r.at = r.at[:0]
+	// convert ts from nano to milli seconds as the iterator work with nanoseconds
+	ts := r.current/1e+6 + r.offset/1e+6
+	for _, series := range r.window {
+		s := r.agg(series.Floats)
+		r.at = append(r.at, promql.Sample{
+			F:      s.F,
+			T:      s.T / int64(time.Millisecond),
+			Metric: series.Metric,
+		})
+	}
+	return ts, SampleVector(r.at)
+}
+
+// The inner batchRangeVectorIterator has preloaded valid samples into each
+// series (see the window filed and load). They're in increasing timestamp
+// order (see next, moves the current timestamp forward by step). So this
+// means that for each downstreamed shard of a first_over_time, selecting the
+// first sample here in each iterator is getting us the earliest timestamped
+// value. Later on when we merge we select the earliest from all the shards.
```

Comment:
I believe, this comment should be broken down into pieces and moved to three different iterators.
`firstWithTimestampBatch`, `lastWithTimestampBatch` and `mergeOverTimeEvaluator`?

Reply:
Done.